### ISSUE TITLE
sockdev device in SDL Hyperion latest causes core dump

### ISSUE THREAD

**Issue by moshix** [SDL-Hercules-390/hyperion#195]
I am using SDL latest.

If I define in VM/370 a device such as `00E 192.168.1.10:1403 ` and then try to connect with it with Fish's Hercprt, it will connect for a split second, then disconnect and cause Hyperion to core dump. Each time, reliably. 

I use Linux 64bit (Ubuntu 18.04) with gcc 7.3.0. 

There is no firewall active, and even if I connect with `telnet 192.168.1.10 1403` from the same Linux host, that will still reliably cause a core dump. 

Thanks.

M

**Comment by moshix** [comment]
Just adding to this that Hercules 3.13 and Hercules 3.12 both on Linux and Windows work fine and don't core dump. seems to be a Hyperion issue. 

thanks

M

**Comment by Fish-Git** [comment]
> If I define in VM/370 a device such as `00E 192.168.1.10:1403`

May I safely presume that you were just typing quickly and what your device statement _actually_ looks like is: `00E  1403  192.168.1.10:1403  sockdev`? (i.e. notice the `1403` device type and the `sockdev` keyword!)

I just tried to repro your crash but was unable to do so. I could connect HercPrt to my Hercules running on my CentOS system just fine. It didn't crash at all.

What would really help us _**a lot**_ when these types of crashes occur, is providing a gdb backtrace so we can know precisely _where_ Hercules crashed. Just saying "Hercules crashes whenever I do xxxx!" isn't all that much help. Especially when we do the same thing and it _doesn't_ crash for us.&nbsp;`;-)`

I'm going to close this as an unresolved "Unknown" until you can provide a gdb backtrace so we can see where it's crashing (which will hopefully tell us _why_ it's crashing too!). If you can provide a backtrace and it looks like it's crashing in Hercules code, then feel free to re-open this issue so we can look into it further.

Until then however, there's no much we can do to help you, so I have no choice (for the time being) but to close this issue as an unresolved Unknown bug.

Thanks.

**Comment by Fish-Git** [comment]
> I am using SDL latest.

For future reference, that doesn't tell us much. It would be better to report the _exact_ version you are using (as reported by the hercules `version` command):

```
08:26:24 HHC01603I version
08:26:24 HHC01413I Hercules version 4.2.0.9576-SDL-ge3c6c3a (4.2.0.9576)
08:26:24 HHC01414I (C) Copyright 1999-2019 by Roger Bowler, Jan Jaeger, and others
08:26:24 HHC01417I ** The SoftDevLabs version of Hercules **
08:26:24 HHC01415I Build date: Feb 28 2019 at 07:57:02
08:26:24 HHC01417I Built with: Clang 3.4.2 (tags/RELEASE_34/dot2-final)
08:26:24 HHC01417I Build type: GNU/Linux x86_64 host architecture build
08:26:24 HHC01417I Modes: S/370 ESA/390 z/Arch
08:26:24 HHC01417I Max CPU Engines: 64
08:26:24 HHC01417I Using   shared libraries
08:26:24 HHC01417I Using   setresuid() for setting privileges
08:26:24 HHC01417I Using   POSIX threads Threading Model
08:26:24 HHC01417I Using   Error-Checking Mutex Locking Model
08:26:24 HHC01417I With    Shared Devices support
08:26:24 HHC01417I With    Dynamic loading support
08:26:24 HHC01417I With    External GUI support
08:26:24 HHC01417I With    IPV6 support
08:26:24 HHC01417I With    HTTP Server support
08:26:24 HHC01417I With    sqrtl support
08:26:24 HHC01417I With    SIGABEND handler
08:26:24 HHC01417I With    CCKD BZIP2 support
08:26:24 HHC01417I With    HET BZIP2 support
08:26:24 HHC01417I With    ZLIB support
08:26:24 HHC01417I With    Regular Expressions support
08:26:24 HHC01417I With    Object REXX support
08:26:24 HHC01417I Without Regina REXX support
08:26:24 HHC01417I With    Automatic Operator support
08:26:24 HHC01417I Without National Language Support
08:26:24 HHC01417I With    CCKD64 Support
08:26:24 HHC01417I Machine dependent assists: cmpxchg1 cmpxchg4 cmpxchg8 hatomics=atomicIntrinsics
08:26:24 HHC01417I Running on: centos-64.localdomain (Linux-2.6.32-754.10.1.el6.centos.plus.x86_64 x86_64) MP=4
08:26:24 HHC01417I Built with decNumber external package version 3.68.0.80-gdb5c456
08:26:24 HHC01417I Built with SoftFloat external package version 3.5.0.83-g3da230f
08:26:24 HHC01417I Built with telnet external package version 1.0.0.42-gcaec0ac
```
&nbsp;
In the above example, the exact reported version is `4.2.0.9576-SDL-ge3c6c3a`.

We don't always need everything else, but also reporting the "Built with:" (`Clang 3.4.2 (tags/RELEASE_34/dot2-final)` and "Build type:" (`GNU/Linux x86_64 host architecture build`) can sometimes help.

The "Running on:" information (`centos-64.localdomain (Linux-2.6.32-754.10.1.el6.centos.plus.x86_64 x86_64) MP=4`) can sometimes help too.

The exact version string however, is the most important. We can use that to match up with the exact repository commit you are using. In the `-ge3c6c3a` portion of the version string, the 7 digits following the `g` are the first 7 digits of the git commit hash, which tells us _exactly_ what code you are using.

Of course, this is all presuming you use git to clone the repository, and then build from that git clone. If you are instead just directly downloading the source .zip file (i.e. _without_ doing an actual git clone of the repository) and building from that, then the reported version number will always be just "4.2.0.0", which doesn't help us much.

**Comment by Fish-Git** [comment]
FYI: &nbsp; it might also help to know the following too:

* Whether or not you tried connecting HercPrt to Hercules _before_ you IPLed your guest, or _after_ it was already up and running.

* What your Hercules configuration file looks like, _especially_ the problematic printer device statement.

* What the IP address is of your Ubuntu _host_. (Remember, HercPrt needs to connect to _Hercules_, not to your guest!)

* What HercPrt parameters you are using (i.e. the IP address and port number you entered into your HercPrt dialog).

* Your Hercules log file (however much of it there is; if Herc crashes there might not be much, but seeing even a _partial_ logfile is better than seeing none at all!)

* Whether you're using any type of unusual routing in your Unbunto host, etc (i.e. anything unusual (out of the ordinary) at all related to networking).

* Anything else you can think of that might possibly be related to the problem.

You mentioned VM/370. Is this the sixpack? The original VM370r6? I'd like to try my test again using the same guest as you are using. Maybe that's where the problem is? Because my test was just a quick one: I started Hercules using the default "sample" configuration file (`hercules.cnf` that comes with the distributed source) but with a socket printer defined at the end, and as I said, it worked fine for me: no crash.

Thanks.

**Comment by moshix** [comment]
See answers below:
Here is the Hercules version

HHC01603I version                                                                                                          
HHC01413I Hercules version 4.2.0.0-SDL-g3a1546c2-modified (4.2.0.0)                                                        
HHC01414I (C) Copyright 1999-2019 by Roger Bowler, Jan Jaeger, and others                                                  
HHC01417I ** The SoftDevLabs version of Hercules **                                                                        
HHC01415I Build date: Feb  4 2019 at 13:12:31                                                                              
HHC01417I Built with: GCC 7.3.0                                                                                            
HHC01417I Build type: GNU/Linux x86_64 host architecture build                                                             
HHC01417I Modes: S/370 ESA/390 z/Arch                                                                                      
HHC01417I Max CPU Engines: 64                                                                                              
HHC01417I Using   shared libraries                                                                                         
HHC01417I Using   setresuid() for setting privileges                                                                       
HHC01417I Using   POSIX threads Threading Model                                                                            
HHC01417I Using   Error-Checking Mutex Locking Model                                                                       
HHC01417I With    Shared Devices support                                                                                   
HHC01417I With    Dynamic loading support                                                                                  
HHC01417I With    External GUI support                                                                                     
HHC01417I With    IPV6 support                                                                                             
HHC01417I With    HTTP Server support                                                                                      
HHC01417I With    sqrtl support                                                                                            
HHC01417I With    SIGABEND handler                                                                                         
HHC01417I With    CCKD BZIP2 support                                                                                       
HHC01417I With    HET BZIP2 support                                                                                        
HHC01417I With    ZLIB support                                                                                             
HHC01417I With    Regular Expressions support                                                                              
HHC01417I Without Object REXX support                                                                                      
HHC01417I Without Regina REXX support                                                                                      
HHC01417I With    Automatic Operator support                                                                               
HHC01417I Without National Language Support                                                                                
HHC01417I With    CCKD64 Support                                                                                           
HHC01417I Machine dependent assists: cmpxchg1 cmpxchg4 cmpxchg8 hatomics=C11                                               
HHC01417I Running on: xps512GB (Linux-4.15.0-45-generic x86_64) MP=4                                                       
HHC01417I Built with decNumber external package version 3.68.0.79-g53f2512                                                 
HHC01417I Built with SoftFloat external package version 3.5.0.82-g1c66591                                                  
HHC01417I Built with telnet external package version 1.0.0.41-ged0ddec            

> 
> * Whether or not you tried connecting HercPrt to Hercules _before_ you IPLed your guest, or _after_ it was already up and running.

I connect with Hercprt right after the IPL of VM/370. I tried to connect before the IPL of VM/370 and same outcome: crash


> * What your Hercules configuration file looks like, _especially_ the problematic printer device statement.

ARCHMODE  S/370
  CPUSERIAL 098052
  CPUMODEL  4381
  LPARNAME  HERCULES
  MAINSIZE  16
  NUMCPU    1
  ECPSVM    NO


  OSTAILOR  QUIET
  SYSEPOCH  1900
  TZOFFSET  -0400
  YROFFSET  0
  PGMPRDOS  RESTRICTED

  PANTITLE  'VM/370 R6 PLC29 Moshix Cloud System'
  PANRATE   SLOW
  CONKPALV  (3,1,10)
  CODEPAGE  819/1047
  TIMERINT  50

  CNSLPORT  3270
  DEVTMAX   0
  SHCMDOPT  NODIAG8

  HERCPRIO    8
  TODPRIO   -20
  DEVPRIO     0
  CPUPRIO     0

  0009  3215-C   NOPROMPT   # 3215 alternate console
# 0010  3270                # 3270 primary console
  000C   3505   3505 sockdev ascii trunc eof
  000D   3525   /dev/null
  000E   1403   192.168.1.34:1403 sockdev
  000F   1403   192.168.1.34:1404 sockdev

#---------------------------------------------------------
#   Direct-access Storage Devices
#---------------------------------------------------------

  # VM/370 system residence volume
  0141   3350   disks/vm3350-1.141.cckd


> * What the IP address is of your Ubuntu _host_. (Remember, HercPrt needs to connect to _Hercules_, not to your guest!)
192.168.1.34

> * What HercPrt parameters you are using (i.e. the IP address and port number you entered into your HercPrt dialog).
I use 192.168.1.34 and port 1403

> * Your Hercules log file (however much of it there is; if Herc crashes there might not be much, but seeing even a _partial_ logfile is better than seeing none at all!)

Here it is:

11:30:19 /Now 13:30:19 GMT THURSDAY 02/28/19
11:30:19 /
11:30:19 /DMKCPI971I System is Uniprocessor generated
11:30:19 /DMKCPI977I Free Trap Installed
11:30:19 /
11:30:19 /DMKUDR476I System Directory loaded from volume VM50-1
11:30:19 /
11:30:19 HHC00107I Starting thread cckd_writer(), active=0, started=0, max=2
11:30:19 HHC00100I Thread id 00007fe47999c700, prio -1, name 'cckd_writer thread 1' started
11:30:19 /13:30:19 AUTO LOGON   ***   OPERATOR USERS = 001  BY  SYSTEM
11:30:19 /13:30:19 DMKCPI954E DASD 6F6 VOLID VSXRES not mounted; duplicate of DASD 149
11:30:19 /
11:30:19 /DMKCPI957I Storage size = 16384 K,   Nucleus = 324 K,
11:30:19 /           Dynamic Paging = 14800 K, Trace Table = 240 K,
11:30:19 /           Free Storage = 1020 K,    Virtual=Real = 00000 K
11:30:19 /
11:30:19 /13:30:19 FILES: 004 RDR,  NO PRT,  NO PUN
11:30:19 /13:30:19 MONITOR AUTO STARTING
11:30:19 /13:30:19 AUTO LOGON   ***   AUTOLOG1 USERS = 002  BY  OPERATOR
11:30:19 /
11:30:19 /DMKCPI966I Initialization complete
11:30:19 /
11:30:20 /13:30:20 RDR  00C STARTED   SYSTEM
11:30:20 /13:30:20 PUN  00D STARTED   SYSTEM   CLASS = P      SEP
11:30:20 /13:30:20 PRT  00E STARTED   SYSTEM   CLASS = A      SEP
11:30:20 /13:30:20 PRT  00F STARTED   SYSTEM   CLASS = A      SEP
11:30:20 /13:30:20 AUTO LOGON   ***   CPWATCH  USERS = 003  BY  AUTOLOG1
11:30:20 /13:30:20 AUTO LOGON   ***   CMSBATCH USERS = 004  BY  AUTOLOG1
11:30:20 /13:30:20 USER DSC LOGOFF AS AUTOLOG1 USERS = 003
11:30:22 HHC01603I version
11:30:22 HHC01413I Hercules version 4.2.0.0-SDL-g3a1546c2-modified (4.2.0.0)
11:30:22 HHC01414I (C) Copyright 1999-2019 by Roger Bowler, Jan Jaeger, and others
11:30:22 HHC01417I ** The SoftDevLabs version of Hercules **
11:30:22 HHC01415I Build date: Feb  4 2019 at 13:12:31
11:30:22 HHC01417I Built with: GCC 7.3.0
11:30:22 HHC01417I Build type: GNU/Linux x86_64 host architecture build
11:30:22 HHC01417I Modes: S/370 ESA/390 z/Arch
11:30:22 HHC01417I Max CPU Engines: 64
11:30:22 HHC01417I Using   shared libraries
11:30:22 HHC01417I Using   setresuid() for setting privileges
11:30:22 HHC01417I Using   POSIX threads Threading Model
11:30:22 HHC01417I Using   Error-Checking Mutex Locking Model
11:30:22 HHC01417I With    Shared Devices support
11:30:22 HHC01417I With    Dynamic loading support
11:30:22 HHC01417I With    External GUI support
11:30:22 HHC01417I With    IPV6 support
11:30:22 HHC01417I With    HTTP Server support
11:30:22 HHC01417I With    sqrtl support
11:30:22 HHC01417I With    SIGABEND handler
11:30:22 HHC01417I With    CCKD BZIP2 support
11:30:22 HHC01417I With    HET BZIP2 support
11:30:22 HHC01417I With    ZLIB support
11:30:22 HHC01417I With    Regular Expressions support
11:30:22 HHC01417I Without Object REXX support
11:30:22 HHC01417I Without Regina REXX support
11:30:22 HHC01417I With    Automatic Operator support
11:30:22 HHC01417I Without National Language Support
11:30:22 HHC01417I With    CCKD64 Support
11:30:22 HHC01417I Machine dependent assists: cmpxchg1 cmpxchg4 cmpxchg8 hatomics=C11
11:30:22 HHC01417I Running on: xps512GB (Linux-4.15.0-45-generic x86_64) MP=4
11:30:22 HHC01417I Built with decNumber external package version 3.68.0.79-g53f2512
11:30:22 HHC01417I Built with SoftFloat external package version 3.5.0.82-g1c66591
11:30:22 HHC01417I Built with telnet external package version 1.0.0.41-ged0ddec
11:31:50 HHC01040I 0:000E COMM: client <unknown>, IP 192.168.1.76 connected to device 192.168.1.34:1403


> * Whether you're using any type of unusual routing in your Unbunto host, etc (i.e. anything unusual (out of the ordinary) at all related to networking).
> * Anything else you can think of that might possibly be related to the problem.

no, none. 

> 
> You mentioned VM/370. Is this the sixpack? The original VM370r6? I'd like to try my test again using the same guest as you are using. Maybe that's where the problem is? Because my test was just a quick one: I started Hercules using the default "sample" configuration file (`hercules.cnf` that comes with the distributed source) but with a socket printer defined at the end, and as I said, it worked fine for me: no crash.

yes, I use Sixpack 1.2
> 
> Thanks.

**Comment by moshix** [comment]
By the way FIsh, I never use a source tar ball or so. I always build from the git repository directly. 

So, when I say SDL latest, I mean the SDL from that very same day when I open up the bug. But yes, I will also provide a gdb backtrace when I get a moment. 

thanks

M

**Comment by Fish-Git** [comment]
> By the way Fish, I never use a source tar ball. I always build from the git repository directly.

Excellent. And I see that in your version information and in your log file too.

> So, when I say SDL latest, I mean the SDL from that very same day when I open up the bug.

_Which,_ according to what you posted just above,  _**is not true!**_ &nbsp; `:(`

Take a close look at your version information, _especially_ the _**build date!**:_

> HHC01413I Hercules version 4.2.0.0-SDL-g3a1546c2-modified (4.2.0.0)                                                        
> HHC01414I (C) Copyright 1999-2019 by Roger Bowler, Jan Jaeger, and others                                                  
> HHC01417I ** The SoftDevLabs version of Hercules **                                                                        
> HHC01415I Build date: Feb  4 2019 at 13:12:31

The precise version you are using (g3a1546c2) corresponds to a version of SDL Hyperion that was [_last updated on **February 3rd!!**_](https://github.com/SDL-Hercules-390/hyperion/commits/master?after=e3c6c3aa173e3a926de9eab67b6a1829a10aaa65+34)

In case you haven't noticed, today is _not_ Feb. 3rd. It's Feb 28th.&nbsp;`;-)`

On [Feb. 9th, Jürgen made a commit](https://github.com/SDL-Hercules-390/hyperion/commits/master) to fix the _exact problem you are currently experiencing!_ ("On certain Linux distributions socket printers segfault when a client connects").

So you aren't running the version of Hercules that you think you are. You are either not refreshing your repository (git pull), or your PATH is messed up causing you to run an older/stale version of Hercules and not the newer one you presumably built, or something else weird is going on.

But th bottom line is, your problem appears to be a bug that was already found and fixed 2.5 weeks ago. Please refresh your repository, rebuild, and retry. I'm sure it'll work much better this time.&nbsp;`:)`

(And _now_ you can see the importance of providing us with the _exact version_ that you are running! So we can catch simple mistakes like this without wasting our collective time chasing down a non-existent bug!)

**Comment by moshix** [comment]
Yes, in this case you're right. I thoght I had updated hyperion on this machine, but I didn't.

glad this is solved. 

thanks a lot

M
